apps/fuzzyw/fuzzyw.app.js

```diff
@@ -31,113 +31,135 @@
     /*LANG*/"ten to *$2",
     /*LANG*/"five to *$2"
   ]
-};
+  };
 
-let text_scale = 4;
-let timeout = 2.5*60;
-let drawTimeout;
-let animInterval;
-let time_string = "";
-let time_string_old = "";
-let time_string_old_wrapped = "";
+  let text_scale = 4;
+  let timeout = 2.5*60;
+  let drawTimeout;
+  let animInterval;
+  let time_string = "";
+  let time_string_old = "";
+  let time_string_old_wrapped = "";
+  let settings = {};
 
-let loadSettings = function() {
-  settings = require("Storage").readJSON(SETTINGS_FILE,1)|| {'showWidgets': false};
-};
+  let loadSettings = function() {
+    settings = require("Storage").readJSON(SETTINGS_FILE,1)|| {'showWidgets': false, 'animate': true};
+  };
 
-let queueDraw = function(seconds) {
-  let millisecs = seconds * 1000;
-  if (drawTimeout) clearTimeout(drawTimeout);
-  drawTimeout = setTimeout(function() {
-    drawTimeout = undefined;
-    draw();
-  }, millisecs - (Date.now() % millisecs));
-};
+  let queueDraw = function(seconds) {
+    let millisecs = seconds * 1000;
+    if (drawTimeout) clearTimeout(drawTimeout);
+    drawTimeout = setTimeout(function() {
+      drawTimeout = undefined;
+      draw();
+    }, millisecs - (Date.now() % millisecs));
+  };
 
-let getTimeString = function(date) {
-  let segment = Math.round((date.getMinutes()*60 + date.getSeconds() + 1)/300);
-  let hour = date.getHours() + Math.floor(segment/12);
-  f_string = fuzzy_string.minutes[segment % 12];
-  if (f_string.includes('$1')) {
-    f_string = f_string.replace('$1', fuzzy_string.hours[(hour) % 12]);
-  } else {
-    f_string = f_string.replace('$2', fuzzy_string.hours[(hour + 1) % 12]);
-  }
-    return f_string;
-};
+  let getTimeString = function(date) {
+    let segment = Math.round((date.getMinutes()*60 + date.getSeconds() + 1)/300);
+    let hour = date.getHours() + Math.floor(segment/12);
+    let f_string = fuzzy_string.minutes[segment % 12];
+    for (let i = 0; i < f_string.length; i++) {
+      if (f_string.charAt(i) == '$') {
+        if (f_string.charAt(i+1) == '1') return f_string.slice(0, i) + fuzzy_string.hours[hour % 12] + f_string.slice(i+2);
+        return f_string.slice(0, i) + fuzzy_string.hours[(hour+1) % 12] + f_string.slice(i+2);
+      }
+    }
+    // the more elegant solution that unfortunately gets optimized to smithereens
```

Comment:
Is this a bug with the pretokenizer then?

Reply:
If I print before the .replace in the original code, I get e.g. "half past *$1" and after I get "half past *", but only with pretokeniser, and not always. Sometimes I get "half past *six". Maybe 50% of the time?

Reply:
Sounds like a possible memory issue in the firmware, perhaps todo with the code that handles template strings? @gfwilliams what do you think?

Reply:
Thanks - yes, this seems to be some firmware issue. I just tried:

```
let f_string = "half past *$1";
print(f_string.replace('$1', "nine"));
```

And when uploaded pretokenised I see the issue, but:

```
let f_string = "half *$1";
print(f_string.replace('$1', "nine"));
```

works (the string is small enough it's loaded to RAM) - I think it's some issue with how `.replace` works on flash-based strings.

Please could you change the code to just do:

```
// add "" to load into RAM due to 2v21 firmware .replace on flashstring issue
f_string = ""+fuzzy_string.minutes[segment % 12]; 
if (f_string.includes('$1')) {
  f_string = f_string.replace('$1', fuzzy_string.hours[(hour) % 12]);
} else {
  f_string = f_string.replace('$2', fuzzy_string.hours[(hour + 1) % 12]);
}
```

That should fix it without you having to add the code to iterate, because adding "" to the string will load it into RAM and then it'll work, and then I'll see what I can do to fix this properly in the firmware

Reply:
This does work for me at least.

Reply:
Ok, just fixed the firmware. So cutting edge or 2v22 (when released) should now have this fixed, but I guess we want to make sure this still works with older firmwares so the `""+` is a good fix

Reply:
Thanks! @thyttan I appreciate you started the review, happy to let you finish it :)

Reply:
Happy it was worked out - thanks!